core-bundle/src/Menu/FrontendMenuBuilder.php

```diff
@@ -0,0 +1,366 @@
+            'showProtected' => false,
+            'showLevel' => 0,
+            'hardLimit' => false,
+            'isSitemap' => false,
+        ], $options);
+
+        if (null === ($pages = $this->getPages($pid, $options))) {
+            return $root;
+        }
+
+        $request = $this->requestStack->getCurrentRequest();
+        /** @var PageModel|null $currentPage */
+        $currentPage = $request->attributes->get('pageModel');
+
+        $isMember = $this->security->isGranted('ROLE_MEMBER');
+
+        /** @var PageModel $page */
+        foreach ($pages as ['page' => $page, 'hasSubpages' => $hasSubpages]) {
+            // Skip hidden sitemap pages
+            if ($options['isSitemap'] && 'map_never' === $page->sitemap) {
+                continue;
+            }
+
+            $page->loadDetails();
+
+            $item = $this->factory->createItem($page->title);
+
+            // Override the domain (see #3765)
+            if (null !== $host) {
+                $page->domain = $host;
+            }
+
+            if ($page->tabindex > 0) {
+                trigger_deprecation('contao/core-bundle', '4.12', 'Using a tabindex value greater than 0 has been deprecated and will no longer work in Contao 5.0.');
+            }
+
+            // Hide the page if it is not protected and only visible to guests (backwards compatibility)
+            if ($page->guests && !$page->protected && $isMember) {
+                trigger_deprecation('contao/core-bundle', '4.12', 'Using the "show to guests only" feature has been deprecated an will no longer work in Contao 5.0. Use the "protect page" function instead.');
+                continue;
+            }
+
+            // PageModel->groups is an array after calling loadDetails()
+            if (
+                $page->protected && !$options['showProtected']
+                && (!($options['isSitemap'] ?? false) || 'map_always' !== $page->sitemap)
+                && !$this->security->isGranted(ContaoCorePermissions::MEMBER_IN_GROUPS, $page->groups)
+            ) {
+                continue;
+            }
+
+            // Check whether there will be subpages
+            if ($hasSubpages) {
+                $nextLevel = $level + 1;
+                $this->getMenu($item, (int) $page->id, $nextLevel, $host, $options);
+
+                $childRecords = $this->database->getChildRecords($page->id, 'tl_page');
+
+                if (
+                    $options['showLevel'] && $options['showLevel'] < $nextLevel
+                    && ($options['hardLimit'] || !$currentPage || ($currentPage->id === $page->id && !\in_array($currentPage->id, $childRecords, false)))
+                ) {
+                    $item->setDisplayChildren(false);
+                }
+            }
+
+            if (null === $href = $this->generateUri($page, $item)) {
+                continue;
+            }
+
+            $this->populateMenuItem($item, $request, $page, $href, $options);
```

Comment:
This should be moved up so that it is excuted before the children are generated. Otherwise `$event->getTree()->getExtras()` will be empty when the `FrontendMenuEvent` is dispatched (since it will be executed depth first) and thus you cannot access the information of the page for the current root node.

Example: depending on a page's setting I want to add new children to the page. Ideally I would implement the following:

```php
namespace App\EventListener;

use Contao\CoreBundle\Event\FrontendMenuEvent;
use Terminal42\ServiceAnnotationBundle\Annotation\ServiceTag;

/**
 * @ServiceTag("kernel.event_listener")
 */
class FrontendMenuListener
{
    public function __invoke(FrontendMenuEvent $event): void
    {
        $tree = $event->getTree();

        // From tl_page.foobar
        if (!$tree->getExtras()['foobar']) {
            return;
        }

        $factory = $event->getFactory();
        $node = $factory->createItem('foobar');
        $tree->addChild($node);
    }
}
```

But in the current state of this PR this would not be possible. Instead I would have to traverse the children each time, which is inefficient.

Reply:
Check out 239a7c024239f274b8e61bea01cd053a57973b58